tracing-core/src/metadata.rs

```diff
@@ -591,6 +555,223 @@ impl fmt::Display for ParseLevelFilterError {
+//    `(HIGHEST_DISCRIMINANT_VALUE + 1)` as the representation for `None`.
+//    Therefore, the integer representation of `LevelFilter::OFF` (which is
+//    `None`) will be the number 5. `OFF` must compare higher than every other
+//    level in order for it to filter as expected. Since we want to use a single
+//    `cmp` instruction, we can't special-case the integer value of `OFF` to
+//    compare higher, as that will generate more code. Instead, we need it to be
+//    on one end of the enum, with `ERROR` on the opposite end, so we assign the
+//    value 0 to `ERROR`.
+//
+//    This *does* mean that when parsing `LevelFilter`s or `Level`s from
+//    `String`s, the integer values are inverted, but that doesn't happen in a
+//    hot path.
+//
+//    Note that we manually invert the comparisons by swapping the left-hand and
+//    right-hand side. Using `Ordering::reverse` generates significantly worse
+//    code (per Matt Godbolt's Compiler Explorer).
+//
+// Anyway, that's a brief history of why this code is the way it is. Don't
+// change it unless you know what you're doing.
+
+impl PartialEq<LevelFilter> for Level {
+    #[inline(always)]
+    fn eq(&self, other: &LevelFilter) -> bool {
+        self.0 as usize == filter_as_usize(&other.0)
+    }
+}
+
+impl PartialOrd for Level {
+    #[inline(always)]
+    fn partial_cmp(&self, other: &Level) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+
+    #[inline(always)]
+    fn lt(&self, other: &Level) -> bool {
+        (other.0 as usize) < (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn le(&self, other: &Level) -> bool {
+        (other.0 as usize) <= (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn gt(&self, other: &Level) -> bool {
+        (other.0 as usize) > (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn ge(&self, other: &Level) -> bool {
+        (other.0 as usize) >= (self.0 as usize)
+    }
+}
+
+impl Ord for Level {
+    #[inline(always)]
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        (other.0 as usize).cmp(&(self.0 as usize))
+    }
+}
+
+impl PartialOrd<LevelFilter> for Level {
+    #[inline(always)]
+    fn partial_cmp(&self, other: &LevelFilter) -> Option<cmp::Ordering> {
+        Some(filter_as_usize(&other.0).cmp(&(self.0 as usize)))
+    }
+
+    #[inline(always)]
+    fn lt(&self, other: &LevelFilter) -> bool {
+        filter_as_usize(&other.0) < (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn le(&self, other: &LevelFilter) -> bool {
+        filter_as_usize(&other.0) <= (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn gt(&self, other: &LevelFilter) -> bool {
+        filter_as_usize(&other.0) > (self.0 as usize)
+    }
+
+    #[inline(always)]
+    fn ge(&self, other: &LevelFilter) -> bool {
+        filter_as_usize(&other.0) >= (self.0 as usize)
+    }
+}
+
+#[inline(always)]
+fn filter_as_usize(x: &Option<Level>) -> usize {
```

Comment:
So... if rustc ever decides to change something in the niche opts (unlikely for this case), this will perf regress. If you want a more reliable check than perf regressions, you could add tests for all 5 cases by transmuting `Option<Level>` to `usize` and comparing it against the result of `filter_as_usize`.

Reply:
Good call, an assertion failure should be much more visible than having to watch the benchmarks, which we don't *currently* gate CI on (it's hard to find a good threshold for what's actually a regression — IMO `criterion` is a bit too sensitive to noise...)